File: 0100-same-tree/0100-same-tree.py
from collections import deque

"""
Understand
Identical = same value, same order of nodes

Match
DFS

Plan
- each time a node is encountered, compare nodes at same position on both trees; 
- if both nodes are Null, return True (reached leaf on both trees, so equal)
- if one node Null and other not Null, return False
- if both nodes not Null AND diff values, return False
- return True
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if (not p or not q) or (q.val != p.val):
            return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: 0100-same-tree/0100-same-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//base case: if both nodes null, true; if values of p and q differ (or one of two is null), false
//recursive step: two trees are same when - cur node is same and left right subtrees are same;
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (!p && !q) {
            return true;
        }
        if ((!p || !q) || (p->val != q->val)) {
            return false;
        }
        
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};